starknet-replay/src/storage/rpc/rpc_client.rs

```diff
@@ -0,0 +1,429 @@
+    /// # Arguments
+    ///
+    /// - `block_number`: the block number at which to query the class hash.
+    /// - `contract_address`: the address of the contract.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err`] if the request fails.
+    #[allow(clippy::missing_panics_doc)] // Needed because `tokio::main` calls `unwrap()`
+    #[tokio::main]
+    pub async fn starknet_get_class_hash_at(
+        &self,
+        block_number: &BlockNumber,
+        contract_address: &ContractAddress,
+    ) -> Result<ClassHash, DatabaseError> {
+        trace!(
+            "starknet_get_class_hash_at {:?} {:?}",
+            block_number,
+            contract_address
+        );
+        let block_id: BlockId = block_number.into();
+        let contract_address: Felt = to_field_element(contract_address);
+        let req = self
+            .get_new_client()
+            .get_class_hash_at(block_id, contract_address)
+            .await;
+        Ok(match req {
+            Ok(class_hash) => Ok(ClassHash(class_hash)),
+            Err(err) => match err {
+                ProviderError::StarknetError(StarknetError::ContractNotFound) => {
+                    Ok(ClassHash(Felt::ZERO))
+                }
+                _ => Err(err),
+            },
+        }?)
+    }
+
+    /// This function queries the value of a storage key.
+    ///
+    /// Returns 0 if the storage key doesn't exist.
+    ///
+    /// # Arguments
+    ///
+    /// - `block_number`: the block number at which to query the storage key.
+    /// - `contract_address`: the address of the contract.
+    /// - `key`: the storage key to query.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err`] if the request fails.
+    #[allow(clippy::missing_panics_doc)] // Needed because `tokio::main` calls `unwrap()`
+    #[tokio::main]
+    pub async fn starknet_get_storage_at(
+        &self,
+        block_number: &BlockNumber,
+        contract_address: &ContractAddress,
+        key: &StorageKey,
+    ) -> Result<Felt, DatabaseError> {
+        trace!(
+            "starknet_get_storage_at {:?} {:?} {:?}",
+            block_number,
+            contract_address,
+            key
+        );
+        let block_id: BlockId = block_number.into();
+        let contract_address: Felt = to_field_element(contract_address);
+        let key: Felt = to_field_element(key);
+        let req = self
+            .get_new_client()
+            .get_storage_at(contract_address, key, block_id)
+            .await;
+        Ok(match req {
+            Ok(storage_value) => Ok(storage_value),
+            Err(err) => match err {
+                ProviderError::StarknetError(StarknetError::ContractNotFound) => Ok(Felt::ZERO),
+                _ => Err(err),
+            },
+        }?)
+    }
+
+    /// This function queries the chain id of the RPC endpoint.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err`] if the request fails or decoding hex values of the chain
+    /// id fails.
```

Comment:
Same comment as above.

Reply:
This method is complex, it can return different types of error.